src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -163,4 +168,147 @@ BOOST_AUTO_TEST_CASE(chainstatemanager_rebalance_caches)
     BOOST_CHECK_CLOSE(c2.m_coinsdb_cache_size_bytes, max_cache * 0.95, 1);
 }
 
+static bool
+CreateAndActivateUTXOSnapshot(NodeContext& node, const fs::path root)
+{
+    // Write out a snapshot to the test's tempdir.
+    //
+    int height;
+    WITH_LOCK(::cs_main, height = node.chainman->ActiveHeight());
+    fs::path snapshot_path = root / tfm::format("test_snapshot.%d.dat", height);
+    FILE* outfile{fsbridge::fopen(snapshot_path, "wb")};
+    CAutoFile auto_outfile{outfile, SER_DISK, CLIENT_VERSION};
+
+    UniValue result = CreateUTXOSnapshot(node, node.chainman->ActiveChainstate(), auto_outfile);
+    BOOST_TEST_MESSAGE(
+        "Wrote UTXO snapshot to " << snapshot_path.make_preferred().string() << ": " << result.write());
+
+    // Read the written snapshot in and then activate it.
+    //
+    FILE* infile{fsbridge::fopen(snapshot_path, "rb")};
+    CAutoFile auto_infile{infile, SER_DISK, CLIENT_VERSION};
+    SnapshotMetadata metadata;
+    auto_infile >> metadata;
+
+    return node.chainman->ActivateSnapshot(auto_infile, metadata, /*in_memory*/ true);
+}
+
+//! Test basic snapshot activation.
+BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100DeterministicSetup)
+{
+    ChainstateManager& chainman = *Assert(m_node.chainman);
+
+    size_t initial_size;
+    size_t initial_total_coins{100};
+
+    // Make some initial assertions about the contents of the chainstate.
+    {
+        LOCK(::cs_main);
+        CCoinsViewCache& ibd_coinscache = chainman.ActiveChainstate().CoinsTip();
+        initial_size = ibd_coinscache.GetCacheSize();
+        size_t total_coins{0};
+
+        for (CTransactionRef& txn : m_coinbase_txns) {
+            COutPoint op{txn->GetHash(), 0};
+            BOOST_CHECK(ibd_coinscache.HaveCoin(op));
+            total_coins++;
+        }
+
+        BOOST_CHECK_EQUAL(total_coins, initial_total_coins);
+        BOOST_CHECK_EQUAL(initial_size, initial_total_coins);
+    }
+
+    // Snapshot should refuse to load at this height.
+    assert(!CreateAndActivateUTXOSnapshot(m_node, m_path_root));
+    BOOST_CHECK(chainman.ActiveChainstate().m_from_snapshot_blockhash.IsNull());
+    BOOST_CHECK_EQUAL(
+        chainman.ActiveChainstate().m_from_snapshot_blockhash,
+        chainman.SnapshotBlockhash().value_or(uint256()));
+
+    // Mine 10 more blocks, putting at us height 110 where a valid assumeutxo value can
+    // be found.
+    mineBlocks(10);
```

Comment:
It's a bit weird to activate a snapshot that is identical to the actual fully validated chain. It would be nice if we could mine headers-only. Calling `ProcessNewBlockHeaders` instead of `ProcessNewBlock` in the test helper might do the trick.

Together with mock time you could test the functionality of waiting for headers too.

Reply:
I don't really understand this. There's no way to generate a snapshot to activate without mining blocks on some chain. The chainstate we're loading the snapshot into has no knowledge of the mined blocks. I don't know how we can do better?